### PR TITLE
Fix sse and nginx buffering

### PR DESCRIPTION
**Pull request by guzzijones**
# Environment:
`PYTHON_VERSION=python3.6 make requirements.txt`

setup nginx `st2.conf` listening on 8443 
```
server {
  listen       *:8443 ssl;
  client_max_body_size 5M;

  ssl on;

  ssl_certificate           /etc/ssl/st2/st2.crt;
  ssl_certificate_key       /etc/ssl/st2/st2.key;
  ssl_session_cache         shared:SSL:10m;
  ssl_session_timeout       5m;

```

client call:
```
python -m st2client.shell --debug  --auth-url https://localhost:8443/auth --api-url https://localhost:8443/api --stream-url https://localhost:8443/stream --cacert=False --skip-config pack install /home/ubuntu/utilstack --python3
```
# What is the problem you're trying to solve with this patch?
https://github.com/StackStorm/st2/issues/4842

# What's the solution?
Solution is to watch the execution id status events and look for one of the LIVEACTION_STATUS_COMPLETED  events

# What's end event and end execution ID do?
tell the `generator` function in stream to stop sending events back to the client

# How is the cacert option has to do with this change? 
allows stream to use the --cacert option via command line.  Needed this to test the change using a self signed certificate.

# Proxy timeout? 
sets a reasonable timeout for connection and for read.  other endpoints have it as well.

### CODE REVIEW

**Comment by guzzijones** [comment]
Well it turns out setting the timeout will cut off the stream going back to the client which causes other issues client side.  Going to have to keep digging here.

File: conf/nginx/st2.conf [hunk 1]
```diff
@@ -85,6 +85,8 @@ server {
     proxy_set_header Host $host;
     proxy_set_header X-Real-IP $remote_addr;
     proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+    proxy_read_timeout 10;
```
Comment (guzzijones) [inline conf/nginx/st2.conf:88]:
```suggestion
    proxy_read_timeout 90;
```

**Comment by guzzijones** [comment]
looks like circle and travis had issues.
afaik this PR should pass.  
Leaving it here for review.

**Comment by guzzijones** [comment]
Looks like this section of code is missing unit testing. I will create a few.

**Review by m4dcoder — CHANGES_REQUESTED** [review]
There's no description for the PR.  It looks like you're trying to fix a hang on your setup.  You have to provide more information in the description.  How is your environment setup?  What is the problem you're trying to solve with this patch?  What's the solution?  What's end event and end execution ID do?  How is the cacert option has to do with this change?  Proxy timeout?  There's a lot of dial that are touched but there's no comments and no description what these options are fixing.

**Comment by guzzijones** [comment]
> Thank you for including detail explanation in the comments of this PR. This goes a long way to help reviewer understand the context of the change(s).
> 
> What happens if something is wrong and the expected end event and end execution ID never arrives? Will the client connection time out? What will the client output looks like in this case?

You would get the same UX as the current implementation of code without this change.  The client will hang on writing the results to stdout and you will need to kill the command with ctrl-c.

**Comment by m4dcoder** [comment]
@guzzijones Thanks for the clarification. Ok with the fix as the events cover most completed statuses and failure case when there's a connection issue or other unforeseen errors.

**Review by m4dcoder — APPROVED** [review]
@guzzijones Thank you for contributing this fix.

**Review by armab — CHANGES_REQUESTED** [review]
Just a reminder that we'll need a Changelog for this fix before merging.

**Comment by guzzijones** [comment]
Ok will do.

File: CHANGELOG.rst [hunk 2]
```diff
@@ -9,6 +9,11 @@ Changed
 * Improve the st2-self-check script to echo to stderr and exit if it isn't run with a
   ST2_AUTH_TOKEN or ST2_API_KEY environment variable. (improvement) #5068
 
+Fixed
+~~~~~~~~~
+* Fix nginx buffering long polling stream to client.  Instead of waiting for closed connection
+  wait for final event to be sent to client. fixes issue #4842
```
Comment (armab) [inline CHANGELOG.rst]:
```suggestion
  wait for final event to be sent to client. (bug fix) #4842  #5042
```
to follow the existing format somewhat closely

Comment (guzzijones) [reply CHANGELOG.rst]: done

**Review by armab — APPROVED** [review]

**Pull request merged**